src/main/scala/com/cognite/sdk/scala/v1/dataModels.scala

```diff
@@ -3,81 +3,104 @@
+    DataModelIdentifier(Some(space), model)
+}
+
 final case class DataModelProperty(
-    `type`: String,
+    `type`: PropertyType,
     nullable: Boolean = true,
-    targetModelExternalId: Option[String] = None
+    targetModel: Option[DataModelIdentifier] = None
 )
 
-final case class DataModelPropertyIndex(
-    indexName: Option[String] = None,
-    fields: Option[Seq[String]] = None
-)
+final case class ContstrainedProperty(property: String)
 
-final case class DataModel(
-    externalId: String,
-    properties: Option[Map[String, DataModelProperty]] = None,
-    `extends`: Option[Seq[String]] = None,
-    indexes: Option[Seq[DataModelPropertyIndex]] = None
+final case class UniquenessConstraint(
+    uniqueProperties: Seq[ContstrainedProperty]
 )
 
-final case class DataModelListInput(includeInheritedProperties: Boolean)
+final case class DataModelConstraints(
+    uniqueness: Option[Map[String, UniquenessConstraint]] = None
+)
 
-final case class DataModelGetByExternalIdsInput[A](
-    items: Seq[A],
-    includeInheritedProperties: Boolean,
-    ignoreUnknownIds: Boolean
+final case class BTreeIndex(
+    properties: Seq[String]
 )
 
-final case class DataModelInstanceCreate(
-    modelExternalId: String,
-    properties: Option[Map[String, PropertyType]] = None
+final case class DataModelIndexes(
+    btreeIndex: Option[Map[String, BTreeIndex]] = None
 )
 
-sealed trait DataModelInstanceFilter
-
-sealed trait DMIBoolFilter extends DataModelInstanceFilter
-final case class DMIAndFilter(and: Seq[DataModelInstanceFilter]) extends DMIBoolFilter
-final case class DMIOrFilter(or: Seq[DataModelInstanceFilter]) extends DMIBoolFilter
-final case class DMINotFilter(not: DataModelInstanceFilter) extends DMIBoolFilter
-
-sealed trait DMILeafFilter extends DataModelInstanceFilter
-final case class DMIEqualsFilter(property: Seq[String], value: PropertyType) extends DMILeafFilter
-final case class DMIInFilter(property: Seq[String], values: Seq[PropertyType]) extends DMILeafFilter
-final case class DMIRangeFilter(
-    property: Seq[String],
-    gte: Option[PropertyType] = None,
-    gt: Option[PropertyType] = None,
-    lte: Option[PropertyType] = None,
-    lt: Option[PropertyType] = None
-) extends DMILeafFilter {
-  require(
-    !(gte.isDefined && gt.isDefined) && // can't have both upper bound in the same time
-      !(lte.isDefined && lt.isDefined) && // can't have both lower bound in the same time
-      (gte.isDefined || gt.isDefined || lte.isDefined || lt.isDefined) // at least one bound must be defined
-  )
+final case class DataModel(
+    externalId: String,
+    properties: Option[Map[String, DataModelProperty]] = None,
+    `extends`: Option[Seq[DataModelIdentifier]] = None,
+    indexes: Option[DataModelIndexes] = None,
+    constraints: Option[DataModelConstraints] = None,
+    instanceType: DataModelInstanceType = DataModelInstanceType.Node
+) {
+  private val (_allowEdge, _allowNode) = instanceType match {
```

Comment:
```suggestion
  private val (allowEdge, allowNode) = instanceType match {
```

Reply:
Scala actually has a built in `Either` type, is that something we could make use of here?

Reply:
For a more "sacala-like" idiom, I went with the `DataModelInstanceType` enumeration that has `Edge` and `Node`, that's what the sdk user will see... but, the api requires those two bools, so they are there only for encoding/decoding

Reply:
Right! It's weird that they designed it this way, when only one can be true, and didn't follow what time series did for `isString`.. which is less clear, but easier to use :shrug:

Reply:
Perhaps we can get rid of the DTO as a whole in the scala sDK by doing explicit encoder/decoder?